backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -54,26 +57,81 @@ public Match findMatch(Participant newParticipant) {
     for (Participant currParticipant : sameDurationParticipants) {
       // Check if participants are both free for that duration + extra
       long currEndTimeAvailable = currParticipant.getEndTimeAvailable();
-      long earliestEndTimeAvailable = Math.min(newEndTimeAvailable, currEndTimeAvailable);
-      boolean compatibleTime =
-          (currentTimeMillis + TimeUnit.MINUTES.toMillis(duration + PADDING_MINUTES))
-              < earliestEndTimeAvailable;
-
-      if (compatibleTime) {
-        // Create and return match
-        return new Match(
-            newParticipant.getUsername(),
-            currParticipant.getUsername(),
-            duration,
-            currentTimeMillis);
+      if (!isCompatibleTime(
+          currentTimeMillis, duration, newEndTimeAvailable, currEndTimeAvailable)) {
+        System.out.println("not compatible time");
+        continue;
       }
+
+      // Check match preference compatibility and get combined preference if compatible
+      MatchPreference currMatchPreference = currParticipant.getMatchPreference();
+      if (!isCompatibleMatchPreference(newMatchPreference, currMatchPreference)) {
+        System.out.println("not compatible match pref");
+        continue;
+      }
+      MatchPreference combinedMatchPreference =
+          getCombinedMatchPreference(newMatchPreference, currMatchPreference);
+
+      if (combinedMatchPreference != MatchPreference.ANY) {
+        int numSameFields = getNumSameFields(newParticipant, currParticipant);
+        System.out.println(numSameFields);
+        if (combinedMatchPreference == MatchPreference.SIMILAR && numSameFields < MIN_SAME_FIELDS) {
+          continue;
+        } else if (combinedMatchPreference == MatchPreference.DIFFERENT
+            && numSameFields >= MIN_SAME_FIELDS) {
+          continue;
+        }
+      }
+      System.out.println("found a match");
+      return new Match(
+          newParticipant.getUsername(), currParticipant.getUsername(), duration, currentTimeMillis);
     }
     // No inital match found
     return null;
   }
+
+  /** Return true if compatible endTimeAvailable considering duration and padding */
+  private boolean isCompatibleTime(
+      long currentTimeMillis, int duration, long newEndTimeAvailable, long currEndTimeAvailable) {
+    long earliestEndTimeAvailable = Math.min(newEndTimeAvailable, currEndTimeAvailable);
+    return (currentTimeMillis + TimeUnit.MINUTES.toMillis(duration + PADDING_MINUTES))
+        < earliestEndTimeAvailable;
+  }
+
+  /** Return true if compatible match preferences (same preference, one or both are any) */
+  private boolean isCompatibleMatchPreference(
+      MatchPreference newMatchPreference, MatchPreference currMatchPreference) {
+    return (newMatchPreference == currMatchPreference)
+        || (newMatchPreference == MatchPreference.ANY)
+        || (currMatchPreference == MatchPreference.ANY);
+  }
```

Comment:
This may produce some unintended results. For example if I ask for a match who is dissimilar to me, and someone very similar to me asks for any match, this logic would allow us to be matched since the other person's match preference is "ANY".

Reply:
This method just says whether their preferences are compatible (only different + similar is not compatible). Then getCombinedMatchPreference finds overall what type of match both would want (similar + any = similar, different + any = different, similar + similar = similar, any + any = any, different + different = different). getSameNumFields says if half or more match and combinedMatchPreference = similar, then match.

Reply:
Not sure if there's a better way to do this, so I'm definitely taking suggestions!

Reply:
Agree with Kevin's comment. What do you think about combining both of the methods into one?

Reply:
Got it. I think the method makes sense as-is but it could use javadoc that includes that explanation. Alternatively Dhruvi's suggestion to merge the methods could also be helpful; the method could return `null` if the two match preferences are incompatible.